source/particles/particle_handler.cc

```diff
@@ -426,6 +426,203 @@ namespace Particles
+    auto n_particles_per_proc =
+      Utilities::MPI::all_gather(triangulation->get_communicator(),
+                                 positions.size());
+
+    // Calculate all starting points locally
+    std::vector<unsigned int> starting_points(n_cpus);
+
+    for (unsigned int i = 0; i < starting_points.size(); ++i)
+      {
+        starting_points[i] = std::accumulate(n_particles_per_proc.begin(),
+                                             n_particles_per_proc.begin() + i,
+                                             0u);
+      }
+
+    auto distributed_tuple =
+      GridTools::distributed_compute_point_locations(cache,
+                                                     positions,
+                                                     global_bounding_boxes);
+
+    // Finally create the particles
+    std::vector<typename Triangulation<dim, spacedim>::active_cell_iterator>
+                                         cell_iterators = std::get<0>(distributed_tuple);
+    std::vector<std::vector<Point<dim>>> dist_reference_points =
+      std::get<1>(distributed_tuple);
+    std::vector<std::vector<unsigned int>> dist_map =
+      std::get<2>(distributed_tuple);
+    std::vector<std::vector<Point<spacedim>>> dist_points =
+      std::get<3>(distributed_tuple);
+    std::vector<std::vector<unsigned int>> dist_procs =
+      std::get<4>(distributed_tuple);
+
+    // Create the multimap of particles
+    std::multimap<typename Triangulation<dim, spacedim>::active_cell_iterator,
+                  Particle<dim, spacedim>>
+      particles;
+
+    // Create the map of cpu to indices, indicating whom sent us what
+    // point
+    std::map<unsigned int, IndexSet> cpu_to_indices;
+
+    for (unsigned int i_cell = 0; i_cell < cell_iterators.size(); ++i_cell)
+      {
+        for (unsigned int i_particle = 0;
+             i_particle < dist_points[i_cell].size();
+             ++i_particle)
+          {
+            const auto &local_id = dist_map[i_cell][i_particle];
+            const auto &cpu      = dist_procs[i_cell][i_particle];
+
+            const unsigned int particle_id = local_id + starting_points[cpu];
+
+            particles.emplace(
+              cell_iterators[i_cell],
+              Particle<dim, spacedim>(dist_points[i_cell][i_particle],
+                                      dist_reference_points[i_cell][i_particle],
+                                      particle_id));
+
+            if (cpu_to_indices.find(cpu) == cpu_to_indices.end())
+              cpu_to_indices.insert({cpu, IndexSet(n_particles_per_proc[cpu])});
+
+            cpu_to_indices[cpu].add_index(local_id);
+          }
+      }
+
+    this->insert_particles(particles);
+    for (auto &c : cpu_to_indices)
+      c.second.compress();
+
+    // Take care of properties, if the input vector contains them.
+    const auto global_n_properties =
+      Utilities::MPI::sum(properties.size(), triangulation->get_communicator());
```

Comment:
It actually does. The second test checks this. If the sum of all properties is zero, then no properties were passed (on no processor). If the sum is not zero, then every one (even when they have no properties locally, because they have no particles locally) need to enter below and exchange properties. The initial assert guarantees that anyone having particles will have also properties (locally with the correct numbering).